core/src/server.cpp

```diff
@@ -104,24 +104,33 @@ void Server::start()
     m_ipban_list = ConfigManager::iprangeBans();
 
     //Rate-Limiter for IC-Chat
-    connect(&next_message_timer, SIGNAL(timeout()), this, SLOT(allowMessage()));
+    connect(&next_message_timer, &QTimer::timeout, this, &Server::allowMessage);
+
+    //Prepare player IDs and reference hash.
+    for (int i = 0; i <= ConfigManager::maxPlayers() - 1; i++){
+        m_available_ids.enqueue(i);
+        m_clients_ids.insert(i, nullptr);
+    }
 }
 
 void Server::clientConnected()
 {
     QTcpSocket* socket = server->nextPendingConnection();
-    int user_id;
-    QList<int> user_ids;
-    for (AOClient* client : qAsConst(m_clients)) {
-        user_ids.append(client->m_id);
-    }
-    for (user_id = 0; user_id <= m_player_count; user_id++) {
-        if (user_ids.contains(user_id))
-            continue;
-        else
-            break;
+
+    //Too many players. Reject connection!
+    //This also enforces the maximum playercount.
+    if (m_available_ids.empty()) {
+        AOPacket disconnect_reason("BD", {"Maximum playercount has been reached."});
```

Comment:
Use `KK` for this, it's misleading to tell the user they've been banned. I know tsu3 does it this way but I've always thought that was dumb.

Reply:
KK would be ignored by the client as the courtroom is not constructed at that point.